implemented toJsonKey to serialize JSON keys

asyncanup opened the pull request:
returns stringified object with keys sorted in alphabetical order

jhusain reviewed (changes requested):
Not convinced the undefined handling is consistent with JSON.stringify at the moment.

Can you add the following tests and confirm they're producing the same output as JSON.stringify?

JSON.stringify({"blah": [undefined,null,undefined]})
"{"blah":[null,null,null]}"

JSON.stringify({"blah": undefined})
"{}"

jhusain commented:
Good idea to add more tests around undefined handling in general, and test them against the behavior of JSON.stringify when possible.

asyncanup commented:
@jhusain fixed.

asyncanup commented:
@jhusain is it looking good now?

lrowe reviewed (changes requested):
As of ES6, [Object key ordering](https://tc39.github.io/ecma262/#sec-ordinary-object-internal-methods-and-internal-slots-ownpropertykeys) is defined and is used by JSON.stringify [of objects](https://tc39.github.io/ecma262/#sec-serializejsonobject). As such, I think it makes sense for us to ensure that the ordering we specify roundtrips through JSON.parse/stringify.

Example:
```js
> ["a", -1, 4, 2, "b"].reduce((acc, key) => { acc[key] = Object.keys(acc).length; return acc; }, {}) 
{ '2': 3, '4': 2, a: 0, '-1': 1, b: 4 }
```

This might be done with the following replacer:
```js
function compare(a, b) {
  const an = Number(a);
  const bn = Number(b);
  const ai = isSafeInteger(an) && String(an) === a && an >= 0;
  const bi = isSafeInteger(bn) && String(bn) === a && bn >= 0;
  if (ai) {
    return bi ? an - bn : -1;
  }
  if (bi) {
    return 1;
  }
  return a < b ? -1 : a === b ? 0 : 1;
}

function replacer(key, value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return value;
  }
  return Object.keys(value)
    .sort(compare)
    .reduce((acc, k) => {
      acc[k] = value[k];
      return acc;
    }, {});
}
```
Though we may also want to order negative integer keys before string keys.

lrowe commented:
Thinking this through we don't even need the `compare` argument to sort in JavaScript since we can rely on the ES6 object ordering to sort integer index keys before string keys:
```js
function replacer(key, value) {
  if (typeof value !== "object" || value === null || Array.isArray(value)) {
    return value;
  }
  return Object.keys(value)
    .sort()
    .reduce((acc, k) => {
      acc[k] = value[k];
      return acc;
    }, {});
}
```

lrowe commented on lib/toJsonKey.js:
```diff
@@ -0,0 +1,49 @@
+        key[0] !== '{' ||
+        key[key.length - 1] !== '}') {
+        return false;
+    }
```
This is just a fast path so might as well just go with the simpler:
```js
    if (typeof key !== 'string' || key[0] !== '{') {
        return false;
    }
```

asyncanup replied:
done

asyncanup commented:
@lrowe fixed up and pushed.

lrowe commented on lib/index.js:
```diff
@@ -3,8 +3,11 @@ module.exports = {
+    isJsonKey: require('./toJsonKey').isJsonKey,
     hasIntersection: require('./hasIntersection'),
     toPaths: require('./toPaths'),
+    isSafeNumber: require('./toPaths').isSafeNumber,
```
Maybe name this `isIntegerKey` for API consistency?

asyncanup replied:
yup, done

lrowe commented on lib/toJsonKey.js:
```diff
@@ -0,0 +1,48 @@
+        obj === null ||
+        Array.isArray(obj)) {
+        return JSON.stringify(obj);
+    }
```
Should we throw a TypeError here since `isJsonKey(result)` will return false?

asyncanup replied:
yes, throwing a TypeError in these cases now, and also throwing a TypeError for other cases such as `Date`, `Map`, `WeakSet`, etc, because `isJsonKey()` will return false for them as well

lrowe commented on lib/toPaths.js:
```diff
@@ -245,4 +245,4 @@ function isSafeNumber(val) {
 
 // export for testing
-module.exports._isSafeNumber = isSafeNumber;
+module.exports.isSafeNumber = isSafeNumber;
```
Let's not change this internal API since falcor-transformer currently depends on it.

asyncanup replied:
I've added it back, but i believe we should make a new PR to falcor-transformer which uses isIntegerKey instead of _isSafeNumber throughout, and then remove _isSafeNumber entirely.

Do you foresee any issues with that? Do any client teams currently use _isSafeNumber?

asyncanup commented:
@lrowe this is up for review once again

asyncanup committed:
implemented toJsonKey to serialize JSON keys

get stringified object with keys sorted in alphabetical order
also added (and exported) isJsonKey to check for valid json keys
added edge-case tests.

lrowe committed:
Break out isIntegerKey / maybeIntegerKey; avoid regex.
